src/Cli/Microsoft.DotNet.Cli.Utils/TypoCorrection.cs

```diff
@@ -2,45 +2,99 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 //
 
-#nullable enable
-
-using System;
 using System.Collections.Generic;
+using System;
 using System.Linq;
 
 namespace Microsoft.DotNet.Cli.Utils
 {
     public static class TypoCorrection
     {
+
+        private enum ComparisonRank { StartsWith = 0, Contains = 1, Levenshtein = 2, NoMatch = 999 }
+
         /// <summary>
-        /// Gets the list of tokens similar to <paramref name="currentToken"/>.
+        /// Gets the list of tokens similar to <paramref name="currentToken"/>
+        /// based on priority search:
+        /// 1. Starts with
+        /// 2. Contains - the call is restricted with <paramref name="currentToken"/> length check, minLength:3 and <param name="maxLevenshteinDistance">
+        /// 3. Levenshtein algorithm with <param name="maxLevenshteinDistance"> restriction
+        /// max number of suggestion is restricted to 10 entries
         /// </summary>
         /// <param name="possibleTokens">List of tokens to select from.</param>
         /// <param name="currentToken">The token that is being compared.</param>
         /// <param name="maxLevenshteinDistance">the difference between two strings, default: 3.</param>
         /// <returns>The enumerator to tokens similar to <paramref name="currentToken"/>.</returns>
-        public static IEnumerable<string> GetSimilarTokens(IEnumerable<string> possibleTokens, string currentToken, int maxLevenshteinDistance = 3)
+        public static IEnumerable<string> GetSimilarTokens(
+            IEnumerable<string> possibleTokens,
+            string currentToken,
+            int maxLevenshteinDistance = 3,
+            int minCurrentTokenLength = 3,
+            int maxNumberOfSuggestions = 10)
         {
+            int suggestionsCounter = 0;
             int? bestDistance = null;
-            return possibleTokens
-                .Select(possibleMatch => (possibleMatch, distance: GetDistance(currentToken, possibleMatch)))
-                .Where(tuple => tuple.distance <= maxLevenshteinDistance)
-                .OrderBy(tuple => tuple.distance)
-                .ThenByDescending(tuple => GetStartsWithDistance(currentToken, tuple.possibleMatch))
-                .TakeWhile(tuple =>
+
+            return possibleTokens.Select(token => GetTokenRank(token, currentToken, minCurrentTokenLength, maxLevenshteinDistance))
+                .Where(rankedToken => rankedToken.Rank != ComparisonRank.NoMatch)
+                .OrderBy(rankedToken => rankedToken.Rank)
+                .ThenBy(rankedToken => rankedToken.Distance)
+                .ThenBy(rankedToken => rankedToken.Rank == ComparisonRank.Levenshtein
+                    ? GetStartsWithDistance(currentToken, rankedToken.Token)
+                    : 0)
```

Comment:
I like the readability and brevity of this code!

Though I must say that I'm unsure that it won't attempt to perform the unwanted sorts. I'd hope so, but quikly peeking into implementation of `ThenBy` and `OrderedEnumerable` it seems to me that it'll do the sort upfront: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs#L21-L32

But I haven't tried - can you try to doublecheck that Levenshtein distance is not counted in case we have enough matches by previous 2 criteria? (just manually create a cese and set breakpoint into the `GetStartsWithDistance`, to verify it's not hit)

Reply:
Jan, I went through the steps and it seems to me that we call `GetDistance` in the `Where` statement when filtering out `rankedToken.Rank != ComparisonRank.NoMatch;`

Reply:
For the sorting, we go down with the evaluated distance for all entries.